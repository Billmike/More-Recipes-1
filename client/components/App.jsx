import React from 'react';
import PropTypes from 'prop-types';
import { Route, Switch, Redirect } from 'react-router-dom';

import HomePage from './homepage';
import CatalogPage from './catalogPage';
import SingleRecipePage from './singleRecipePage';
import UserDashboardPage from './userRecipes';
import NotFoundPage from './notFoundPage';
import { decodeToken } from '../utils/authenticate';
import '../public/style.scss';

const propTypes = {
  location: PropTypes.shape(),
};

const defaultProps = {
  location: undefined,
};

/**
 * App component
 *
 * @returns {any} - Navigate to the matched route
 */
const App = () => (
  <Switch>
    <Route exact path="/" component={HomePage} />
    <Route exact path="/catalog" component={CatalogPage} />
    <Route exact path="/recipes/:recipeId" component={SingleRecipePage} />
    <Route
      path="/users/:userId/dashboard"
      render={props => (decodeToken()
        ? <UserDashboardPage {...props} />
        : <Redirect to={{
          pathname: '/',
          state: {
            from: props.location
          }
        }}
        />)
      }
    />
    <Route component={NotFoundPage} />
  </Switch>
);

App.propTypes = propTypes;
App.defaultProps = defaultProps;

export default App;
