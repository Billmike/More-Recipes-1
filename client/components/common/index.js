import Header from './Header';
import Footer from './Footer';
import RecipeImage from './RecipeImage';

export default {
  Header,
  Footer,
  RecipeImage,
};
