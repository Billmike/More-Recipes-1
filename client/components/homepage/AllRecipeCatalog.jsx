import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import RecipeCatalogCard from '../common/RecipeCatalogCard';

const propTypes = {
  recipesCatalog: PropTypes.arrayOf(PropTypes.shape()),
};

const defaultProps = {
  recipesCatalog: undefined,
};

/**
 * All recipes catalog component
 *
 * @param {object} props - The properties passsed to the component
 *
 * @returns {object} React element
 */
const AllRecipeCatalog = (props) => {
  const { recipesCatalog } = props;

  return (
    <div className="row">
      <div className="col s12" >
        <div className="row valign-wrapper" >
          <h4 className="col s11">All Recipes</h4>
          <span className="col s1 right" >
            <Link to="/catalog">View All &gt;</Link>
          </span>
        </div>
      </div>
      {recipesCatalog && recipesCatalog.map((recipe, index) =>
        (<RecipeCatalogCard
          {...props}
          key={recipe.id}
          index={index}
          recipe={recipe}
        />))}
    </div>
  );
};

AllRecipeCatalog.propTypes = propTypes;
AllRecipeCatalog.defaultProps = defaultProps;

export default AllRecipeCatalog;
