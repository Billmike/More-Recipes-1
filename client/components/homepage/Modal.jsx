import React from 'react';
import PropTypes from 'prop-types';
import Dialog from 'material-ui/Dialog';
import FlatButton from 'material-ui/FlatButton';

import Signup from './Signup';

const customContentStyle = {
  width: '35%',
  maxWidth: '35%',
};

const Modal = (props) => {
  const actions = [
    <FlatButton
      label="Cancel"
      secondary
      onClick={props.handleClose}
    />
  ];

  return (
    <div>
      <Dialog
        title="Create an account"
        actions={actions}
        modal
        contentStyle={customContentStyle}
        open={props.open}
        autoScrollBodyContent
      >
        <Signup />
      </Dialog>
    </div>
  );
};

Modal.propTypes = {
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired
};

export default Modal;
