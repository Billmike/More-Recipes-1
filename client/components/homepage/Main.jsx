import React from 'react';
import PropTypes from 'prop-types';

import PopularRecipesCatalog from './PopularRecipesCatalog';
import AllRecipeCatalog from './AllRecipeCatalog';
import SearchBar from '../common/SearchBar';
import Slider from './Slider';

const propTypes = {
  recipes: PropTypes.shape({
    recipes: PropTypes.arrayOf(PropTypes.shape()),
    popularRecipes: PropTypes.arrayOf(PropTypes.shape()),
  }),
};

const defaultProps = {
  recipes: undefined,
};

/**
 * Main component
 *
 * @param {object} props - The properties passed to the component
 *
 * @returns {object} React element
 */
const Main = (props) => {
  const { recipes: { recipes, popularRecipes } } = props;

  return (
    <div>
      <div id="position">
        <Slider />
        <SearchBar
          id={'home-search-bar'}
          {...props}
        />
      </div>
      <div className="grey lighten-3" >
        <div className="row recipes-list">
          <div className="col s12 popular-list white">
            <PopularRecipesCatalog popularRecipes={popularRecipes} {...props} />
          </div>
          <div className="col s12 all-recipes-list white">
            <AllRecipeCatalog recipesCatalog={recipes} />
          </div>
        </div>
      </div>
    </div>
  );
};

Main.propTypes = propTypes;
Main.defaultProps = defaultProps;

export default Main;
