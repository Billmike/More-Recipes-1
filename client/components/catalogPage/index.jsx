import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import actionCreators from '../../actions';
import Header from '../common/Header';
import Main from './Main';
import Footer from '../common/Footer';

const propTypes = {
  searchRecipe: PropTypes.func.isRequired,
  location: PropTypes.shape({
    state: PropTypes.shape({
      search: PropTypes.string,
      searchTerm: PropTypes.string,
    }),
  }),
  recipes: PropTypes.shape({
    searchResult: PropTypes.arrayOf(PropTypes.shape()),
  }).isRequired,
};

const defaultProps = {
  location: undefined,
};

/**
 * Class representing catalog page
 *
 * @class CatalogPage
 * @extends {Component}
 */
class CatalogPage extends Component {
  /**
   * Creates an instance of CatalogPage.
   *
   * @memberof CatalogPage
   */
  constructor() {
    super();
    this.state = {
      searchBy: 'name',
      searchTerm: '',
      searchedTerm: '',
      showText: false,
    };
  }

  /**
   * Component did mount lifecycle method
   *
   * @returns {any} Any
   * @memberof CatalogPage
   */
  componentDidMount() {
    const { location: { state }, searchRecipe } = this.props;
    window.scrollTo(0, 0);
    if (state) {
      searchRecipe(state.searchBy, state.searchTerm);
    }
  }

  /**
   * Component will receive props lifecycle method
   *
   * @returns {any} Sets state
   * @memberof CatalogPage
   */
  componentWillReceiveProps() {
    if (this.props.recipes.searchResult) {
      this.setState({
        showText: true,
      });
    }
  }

  /**
   * Function to handle change in search term
   *
   * @param {object} event - The event object
   *
   * @returns {any} Changes state to the current search term
   * @memberof CatalogPage
   */
  handleSearchChange = (event) => {
    const { target: { value } } = event;

    if (value === 'name') {
      this.setState(() => ({
        searchBy: value
      }));
    } else if (value === 'ingredients') {
      this.setState(() => ({
        searchBy: value,
      }));
    } else {
      this.setState(() => ({
        searchTerm: value
      }));
    }
  }

  /**
   * Function to handle submitting search term
   *
   * @param {object} event - The event object
   *
   * @returns {any} Submits the search term
   * @memberof CatalogPage
   */
  handleSubmitSearch = (event) => {
    event.preventDefault();
    const { searchBy, searchTerm } = this.state;
    const { searchRecipe } = this.props;

    this.setState(() => ({
      searchedTerm: searchTerm,
    }));

    searchRecipe(searchBy, searchTerm);
  }

  /**
   * Render method
   *
   * @returns {object} React element
   * @memberof CatalogPage
   */
  render() {
    return (
      <div>
        <header>
          <Header {...this.props} />
        </header>
        <main>
          <Main
            handleSearchChange={this.handleSearchChange}
            handleSubmitSearch={this.handleSubmitSearch}
            {...this.state}
            {...this.props}
          />
        </main>
        <footer>
          <Footer />
        </footer>

      </div>
    );
  }
}

/**
 * Function to map values from state to props
 *
 * @param {any} state - The state values
 *
 * @returns {object} - The mapped props
 */
const mapStateToProps = state => ({
  recipes: state.recipes,
  user: state.user,
});

/**
 * Function to map dispatch to props
 * Action creators are binded to the dispatch function
 *
 * @param {any} dispatch - The store dispatch function
 *
 * @returns {any} The mapped props
 */
const mapDispatchToProps = dispatch => (
  bindActionCreators(actionCreators, dispatch)
);

CatalogPage.propTypes = propTypes;
CatalogPage.defaultProps = defaultProps;

export default connect(mapStateToProps, mapDispatchToProps)(CatalogPage);
