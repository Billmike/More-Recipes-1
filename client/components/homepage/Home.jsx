import React from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import actionCreators from '../../actions';
import common from '../common';
import Main from './Main';

const { Header, Footer } = common;

/**
 * Class representing the home component
 *
 * @class Home
 * @extends {React.Component}
 */
class Home extends React.Component {
  /**
   * Component did mount function
   *
   * @returns {object} The response from the server
   * @memberof Home
   */
  componentDidMount() {
    this.props.retrieveRecipes();
  }

  /**
   * The render function
   *
   * @returns {any} React element
   * @memberof Home
   */
  render() {
    return (
      <div>
        <Header {...this.props} />
        <Main {...this.props} />
        <Footer />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  recipes: state.recipes,
  user: state.user,
});

const mapDispatchToProps = dispatch => (
  bindActionCreators(actionCreators, dispatch)
);

Home.propTypes = {
  retrieveRecipes: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(Home);
