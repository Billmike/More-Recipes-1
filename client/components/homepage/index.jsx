import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import actionCreators from '../../actions';
import Header from '../common/Header';
import Footer from '../common/Footer';
import Main from './Main';

const propTypes = {
  recipes: PropTypes.shape({
    recipes: PropTypes.arrayOf(PropTypes.shape()),
  }).isRequired,
  retrieveRecipes: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func,
  }).isRequired,
};

/**
 * Class representing the home component
 *
 * @class Home
 *
 * @extends {Component}
 */
class Home extends Component {
  /**
   * Creates an instance of Home.
   *
   * @memberof Home
   */
  constructor() {
    super();
    this.state = {
      search: 'title',
      list: '',
    };
  }

  /**
   * Component did mount function
   *
   * @returns {object} The response from the server
   *
   * @memberof Home
   */
  componentDidMount() {
    const { recipes: { recipes } } = this.props;

    if (!recipes.length > 0) {
      this.props.retrieveRecipes();
    }
  }

  /**
   * Function to handle change in search term
   *
   * @param {object} event - The event object
   *
   * @returns {any} Changes state to the current search term
   *
   * @memberof Home
   */
  handleSearchChange = (event) => {
    const { target: { value } } = event;

    if (value === 'title') {
      this.setState(() => ({
        search: value
      }));
    } else if (value === 'ingredients') {
      this.setState(() => ({
        search: value,
      }));
    } else {
      this.setState(() => ({
        list: value
      }));
    }
  }

  /**
   * Function to handle submitting search term
   *
   * @param {object} event - The event object
   *
   * @returns {any} Submits the search term
   *
   * @memberof Home
   */
  handleSubmitSearch = (event) => {
    event.preventDefault();
    const { search, list } = this.state;

    if (list !== '') {
      this.props.history.push({
        pathname: '/catalog',
        search: `?search=${search}&list=${list}`,
      });
    }
  }

  /**
   * The render function
   *
   * @returns {object} React element
   *
   * @memberof Home
   */
  render() {
    return (
      <div className="page-body">
        <header>
          <Header {...this.props} />
        </header>
        <main>
          <Main
            handleSearchChange={this.handleSearchChange}
            handleSubmitSearch={this.handleSubmitSearch}
            {...this.state}
            {...this.props}
          />
        </main>
        <footer>
          <Footer />
        </footer>
      </div>
    );
  }
}

/**
 * Function to map values from state to props
 *
 * @param {any} state - The state values
 *
 * @returns {object} - The mapped props
 */
const mapStateToProps = state => ({
  recipes: state.recipes,
  user: state.user,
  userRecipes: state.userRecipes,
});

/**
 * Function to map dispatch to props
 * Action creators are binded to the dispatch function
 *
 * @param {any} dispatch - The store dispatch function
 *
 * @returns {any} The mapped props
 */
const mapDispatchToProps = dispatch => (
  bindActionCreators(actionCreators, dispatch)
);

Home.propTypes = propTypes;

export default connect(mapStateToProps, mapDispatchToProps)(Home);
