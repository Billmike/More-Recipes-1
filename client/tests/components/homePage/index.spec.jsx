import React from 'react';

import { Home } from '../../../components/homepage';

describe('Home component', () => {
  it('should render component', () => {
    const props = {
      recipes: {
        recipes: [],
      },
      retrievePopularRecipes: jest.fn(),
      retrieveRecipes: jest.fn(),
      history: {
        push: jest.fn(),
      },
      location: {},
      resetAuthentication: jest.fn(),
    };
    const wrapper = shallow(<Home {...props} />);

    expect(wrapper).toBeDefined();
  });
  it('should notify users to login when redirected to home', () => {
    const props = {
      recipes: {
        recipes: [],
      },
      retrievePopularRecipes: jest.fn(),
      retrieveRecipes: jest.fn(),
      history: {
        push: jest.fn(),
      },
      location: {
        state: {
          message: 'Unauthenticated user',
        }
      },
      resetAuthentication: jest.fn(),
    };
    const wrapper = shallow(<Home {...props} />);

    expect(wrapper).toBeDefined();
    expect(props.resetAuthentication).toHaveBeenCalled();
  });
  it('should simulate calling class instance properties', () => {
    const props = {
      recipes: {
        recipes: [],
      },
      retrievePopularRecipes: jest.fn(),
      retrieveRecipes: jest.fn(),
      history: {
        push: jest.fn(),
      },
      location: {
        state: {
          message: 'Unauthenticated user',
        }
      },
      resetAuthentication: jest.fn(),
    };
    const event = {
      preventDefault: jest.fn(),
    };
    const wrapper = shallow(<Home {...props} />);

    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
    wrapper.instance().handleSearchChange({ target: { value: 'title' } });
    expect(wrapper.instance().state.search).toEqual('title');
    wrapper.instance().handleSearchChange({ target: { value: 'ingredients' } });
    expect(wrapper.instance().state.search).toEqual('ingredients');
    wrapper.instance().handleSearchChange(
      { target: { value: 'rice, tomatoes' } }
    );
    expect(wrapper.instance().state.search).toEqual('ingredients');
    expect(wrapper.instance().state.list).toEqual('rice, tomatoes');
    expect(wrapper).toMatchSnapshot();
    wrapper.instance().handleSubmitSearch(event);
    expect(props.history.push).toBeCalled();
    wrapper.instance().handleSearchCategory({ target: { name: 'breakfast' } });
    expect(props.history.push).toBeCalled();
    expect(wrapper).toMatchSnapshot();
  });
});

