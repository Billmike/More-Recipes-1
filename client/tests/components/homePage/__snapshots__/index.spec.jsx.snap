// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component should simulate calling class instance properties 1`] = `
<div
  className="page-body"
>
  <header>
    <Header
      handleSearchCategory={[Function]}
      history={
        Object {
          "push": [MockFunction],
        }
      }
      location={
        Object {
          "state": Object {
            "message": "Unauthenticated user",
          },
        }
      }
      recipes={
        Object {
          "recipes": Array [],
        }
      }
      resetAuthentication={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrievePopularRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
    />
  </header>
  <main>
    <Main
      handleSearchChange={[Function]}
      handleSubmitSearch={[Function]}
      history={
        Object {
          "push": [MockFunction],
        }
      }
      list=""
      location={
        Object {
          "state": Object {
            "message": "Unauthenticated user",
          },
        }
      }
      recipes={
        Object {
          "recipes": Array [],
        }
      }
      resetAuthentication={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrievePopularRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      search="title"
    />
  </main>
  <footer>
    <Footer />
  </footer>
</div>
`;

exports[`Home component should simulate calling class instance properties 2`] = `
<div
  className="page-body"
>
  <header>
    <Header
      handleSearchCategory={[Function]}
      history={
        Object {
          "push": [MockFunction],
        }
      }
      location={
        Object {
          "state": Object {
            "message": "Unauthenticated user",
          },
        }
      }
      recipes={
        Object {
          "recipes": Array [],
        }
      }
      resetAuthentication={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrievePopularRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
    />
  </header>
  <main>
    <Main
      handleSearchChange={[Function]}
      handleSubmitSearch={[Function]}
      history={
        Object {
          "push": [MockFunction],
        }
      }
      list=""
      location={
        Object {
          "state": Object {
            "message": "Unauthenticated user",
          },
        }
      }
      recipes={
        Object {
          "recipes": Array [],
        }
      }
      resetAuthentication={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrievePopularRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      search="title"
    />
  </main>
  <footer>
    <Footer />
  </footer>
</div>
`;

exports[`Home component should simulate calling class instance properties 3`] = `
<div
  className="page-body"
>
  <header>
    <Header
      handleSearchCategory={[Function]}
      history={
        Object {
          "push": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "pathname": "/catalog",
                  "search": "?search=ingredients&list=rice, tomatoes",
                },
              ],
              Array [
                Object {
                  "pathname": "/catalog",
                  "search": "?search=category&list=breakfast",
                },
              ],
            ],
          },
        }
      }
      location={
        Object {
          "state": Object {
            "message": "Unauthenticated user",
          },
        }
      }
      recipes={
        Object {
          "recipes": Array [],
        }
      }
      resetAuthentication={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrievePopularRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
    />
  </header>
  <main>
    <Main
      handleSearchChange={[Function]}
      handleSubmitSearch={[Function]}
      history={
        Object {
          "push": [MockFunction] {
            "calls": Array [
              Array [
                Object {
                  "pathname": "/catalog",
                  "search": "?search=ingredients&list=rice, tomatoes",
                },
              ],
              Array [
                Object {
                  "pathname": "/catalog",
                  "search": "?search=category&list=breakfast",
                },
              ],
            ],
          },
        }
      }
      list=""
      location={
        Object {
          "state": Object {
            "message": "Unauthenticated user",
          },
        }
      }
      recipes={
        Object {
          "recipes": Array [],
        }
      }
      resetAuthentication={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrievePopularRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      retrieveRecipes={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      search="title"
    />
  </main>
  <footer>
    <Footer />
  </footer>
</div>
`;
